README
x360: A set of tools for working with the Xbox 360
Copyright (C) 2009-2011 by Isaac Tepper
Licensed under the terms of the GNU GPL v3


CONTENTS
========

1. Introduction
2. Building
3. Tools
   * xpart: Xbox 360 Partition Helper
   * xfd: Xbox Filesystem Driver
4. Libraries
   * libfatx: Userspace access to a FATX filesystem



INTRODUCTION
============

x360 is a set of tools that allow linux users to interface with their
Xbox 360. Currently, it is comprised of a filesystem driver and a
matching library (and a partition helper), but it will soon have some
other tools, such as a gamesave resigner and a media manager.

The name "x360" used to refer directly to the filesystem driver, but
now it refers to the collection of tools in general.



BUILDING
========

x360 has moved over to an automake build system. To build, simply run

     ./configure
     make
     make install

in the project directory. You can type "./configure --help" for more
information.



TOOLS
=====

xpart (Xbox 360 Partition Helper)

Usage: xpart /dev/sdX

Purpose: Splits /dev/sdX into three virtual device files representing
the three (known) partitions: /dev/sdx1, /dev/sdx2, and
/dev/sdx3. sdx1 will be the cache partition, sdx2 will be the Xbox
compatability partition, and sdx3 will be the Xbox 360 content
partition.

Note: There are two versions: xpart-dm, which uses device manager,
and xpart-loop, which uses loop devices. By default, xpart will be a
symlink to xpart-dm.



xfd (Xbox Filesystem Driver)

Usage: xfd-mount [options] /dev/sdcX /mnt
       mount -t fatx [options] /dev/sdcX /mnt

Options are: -d: debug mode, fuse does not daemonize and prints
syscalls as they happen. Also useful for getting warnings from
libfatx.

Purpose: Mounts a FATX partition, allowing you to read and change it's
contents (but currently libfatx has read support only). xfd (or, more
specifically, libfatx) has support for filesystems of any size, big or
little endian, with 16 bit FAT or 32 bit FAT.

Note: You can invoke this through "mount -t fatx", but that method
requires you to be root, whereas simply using xfd-mount does not
require you to be root. Not being root is always better :)



LIBRARIES
=========

libfatx (Userspace access to a FATX filesystem)

libfatx handles the filesystem access for xfd, and will for other
utilities in the future (such as a media manager). Like everything
else, it is a work in progress, and this list of features will
describe its current state:

Implemented Features:

* Runtime detection and handling of little-endian/big-endian
  filesystems

* Runtime detection and handling of 16-bit/32-bit filesystems

* Directory listing that steps through multi-cluster directories

* File/directory offset locator that steps through multi-cluster
  directories

* Reading files

* Translating FATX timestamps to/from unix time

Features Coming Soon:

* Write files, truncate (expand and shrink) files, create files,
  delete files

libfatx is a complete re-write, which is why write support is not
implemented yet. It's been thoroughly tested (as opposed to the
previous version which appearantly froze up a lot), and everything
that is implemented should work without error.

NEWS
Sample NEWS file for x360 project.


x360.h
#ifndef __X360_H
#define __X360_H

#define FUSE_USE_VERSION 26
#define _GNU_SOURCE

#include <fuse.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <sys/types.h>

typedef struct {
    off_t start;
    off_t fat;
    off_t root_dir;
    size_t size;
    uint32_t fat_num;
    uint32_t *fat_ptr;
} x360_partition;

typedef struct {
    uint32_t magic;
    uint32_t vol_id;
    uint32_t cl_size;
    uint32_t fat_num;
    uint16_t zero;
}__attribute__((packed)) x360_boot_sector;

#define X360_FAT_MAX 0x6B0800
typedef uint32_t x360_fat[X360_FAT_MAX];

typedef struct {
    uint8_t fnsize;
    uint8_t attribute;
    int8_t filename[42];
    uint32_t cluster;
    uint32_t fsize;
    uint16_t mtime;
    uint16_t mdate;
    uint16_t ctime;
    uint16_t cdate;
    uint16_t atime;
    uint16_t adate;
}__attribute__((packed)) x360_file_record;

// x360_fr.c
int x360_fr_create(x360_file_record *fr, void *data);
int x360_fr_mkdir(x360_file_record *fr, void *data);
int x360_fr_rename(x360_file_record *fr, void *data);
int x360_fr_truncate(x360_file_record *fr, void *data);
int x360_fr_unlink(x360_file_record *fr, void *data);

#define X360_DIR_MAX 256
typedef x360_file_record x360_dir_cluster[X360_DIR_MAX];

typedef int (*x360_function) (x360_file_record *fr, void *data);

#endif


x360_fr.c
#include "x360.h"

extern x360_partition pt;

int x360_fr_create(x360_file_record *fr, void *data) {
    int ret;
    uint32_t cl;
    memset(fr, 0, sizeof(x360_file_record));
    ret = x360_fr_rename(fr, data);
    if (ret < 0) return ret;
    cl = x360_find_free_cluster();
    if (cl == 0) return -ENOSPC;
    fr->cluster = __bswap_32(cl);
    pt.fat_ptr[cl] = 0xFFFFFFFF;
    return 0;
}

int x360_fr_mkdir(x360_file_record *fr, void *data) {
    int ret;
    ret = x360_fr_create(fr, data);
    if (ret < 0) return ret;
    fr->attribute |= 0x10;
    return 0;
}

int x360_fr_rename(x360_file_record *fr, void *data) {
    char *new = (char *) data;
    char *base = strdupa(new);
    base = basename(base);
    size_t len = strlen(base);
    if (len > 42) return -ENAMETOOLONG;
    fr->fnsize = len;
    memset(fr->filename, 0xFF, 42);
    memcpy(fr->filename, base, len);
    return 0;
}

int x360_fr_truncate(x360_file_record *fr, void *data) {
    off_t *size = (off_t *) data;
    uint32_t cl, cl2;
    uint32_t max = x360_clusters(fr) << 14;

    while (*size > max) {
        cl = x360_cluster(fr, (max >> 14) - 1);
        cl2 = x360_find_free_cluster();
        if (cl2 == 0) return -ENOSPC;
        pt.fat_ptr[cl] = __bswap_32(cl2);
        max += 0x4000;
    }

    while (*size <= (max - 0x4000)) {
        if (*size == 0) break;
        cl = x360_cluster(fr, (max >> 14) - 1);
        if (cl == 0) return -EIO;
        pt.fat_ptr[cl] = 0;
        max -= 0x4000;
    }

    cl = x360_cluster(fr, (max >> 14) - 1);
    pt.fat_ptr[cl] = 0xFFFFFFFF;
    fr->fsize = __bswap_32(*size);
    return 0;
}

int x360_fr_unlink(x360_file_record *fr, void *data) {
    (void) data;
    int32_t cl, cl2;
    fr->fnsize = 0xE5;
    cl = __bswap_32(fr->cluster);
    while ((cl & 0xFFFFFFF) != 0xFFFFFFF) {
        cl2 = __bswap_32(pt.fat_ptr[cl]);
        pt.fat_ptr[cl] = 0;
        cl = cl2;
    }
    return 0;
}


src/xfd/xfd.c
/*
  xfd: FATX filesystem driver
  Copyright (C) 2010-2011  Isaac Tepper <Isaac356@live.com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#define FUSE_USE_VERSION 26

#include "fatx.h"
#include <fuse.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>

static fatx_fs_info *info;

static struct readdir_user {
	fuse_fill_dir_t filler;
	void *buf;
};

static int xfd_getattr(const char *path, struct stat *stbuf)
{
    int res = 0;
    fatx_file_record record;

    memset(stbuf, 0, sizeof(struct stat));

    res = fatx_read_file_record(&record, info, path);
    if (res < 0) return res;

    if (record.isdir) {
    	stbuf->st_mode = S_IFDIR|0755;
    	stbuf->st_nlink = 2;
    } else {
    	stbuf->st_mode = S_IFREG|0644;
    	stbuf->st_nlink = 1;
    	stbuf->st_size = record.size;
    }
    stbuf->st_uid = getuid();
    stbuf->st_gid = getgid();
    stbuf->st_ctim.tv_sec = record.created;
    stbuf->st_mtim.tv_sec = record.modified;
    stbuf->st_atim.tv_sec = record.accessed;

    return res;
}

static void readdir_callback(const char *path, struct readdir_user *user) {
	user->filler(user->buf, path, NULL, 0);
}

static int xfd_readdir(const char *path, void *buf, fuse_fill_dir_t filler,
                         off_t offset, struct fuse_file_info *fi)
{
    (void) offset;
    (void) fi;

    struct readdir_user user = {
    		.filler = filler,
    		.buf = buf
    };

    filler(buf, ".", NULL, 0);
    filler(buf, "..", NULL, 0);

    fatx_list_dir(info, path, readdir_callback, &user);

    return 0;
}

static int xfd_read(const char *path, char *buf, size_t size, off_t offset,
                      struct fuse_file_info *fi)
{
    size_t len;

    len = fatx_read_file(info, path, buf, size, offset);

    return len;
}

static struct fuse_operations xfd_oper = {
    .getattr	= xfd_getattr,
    .readdir	= xfd_readdir,
    .read	= xfd_read
};

int main(int argc, char *argv[])
{
	int debug, fargc, c;
	debug = 0;
	while ((c = getopt(argc, argv, "d")) != -1) {
		switch (c) {
		case 'd':
			debug = 1;
			break;
		}
	}
	fargc = debug ? 3 : 2;
	char *fargv[4] = {argv[0], argv[optind + 1], debug ? "-d" : NULL, NULL};
	info = fatx_fs_init(argv[optind]);
	if (info == NULL) return -1;
    int ret = fuse_main(fargc, fargv, &xfd_oper, NULL);
    fatx_fs_end(info);
    return ret;
}


trunk/src/include/fatx.h
/*
 * libfatx.h
 *
 *  Created on: Dec 15, 2010
 *      Author: ice
 */

#ifndef FATX_H_
#define FATX_H_

#include <stdio.h>
#include <stddef.h>
#include <time.h>

typedef struct fatx_fs_info fatx_fs_info;

typedef struct fatx_file_record {
	char name[43];
	size_t size;
	time_t modified;
	time_t created;
	time_t accessed;
	int isdir;
} fatx_file_record;

typedef struct fatx_file_offsets {
	off_t record_offset;
	off_t data_offset;
} fatx_file_offsets;

fatx_fs_info *fatx_fs_init(const char *filename);
void fatx_fs_end(fatx_fs_info *info);
int fatx_find_file_offsets(struct fatx_file_offsets *offsets,
		fatx_fs_info *info, const char *path);
int fatx_read_file_record(fatx_file_record *file_record,
		fatx_fs_info *info, const char *path);
int fatx_list_dir(fatx_fs_info *info, const char *path, void (*func)(const char *, void *), void *user);

#endif /* FATX_H_ */


trunk/src/libfatx/fatx.c
/*
  libfatx: Userspace access to a FATX filesystem
  Copyright (C) 2010  Isaac Tepper <Isaac356@live.com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <fatx.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <fcntl.h>
#include <unistd.h>
#include <time.h>
#include <libgen.h>
#include <endian.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <errno.h>
#include <string.h>

#define FATX_MAGIC 0x46415458
#define max(a, b) (((a) > (b)) ? (a) : (b))

const char *delimiter = "/";

struct fatx_fs_info {
	int fd;
	int endianness;
	size_t width;
	int mode;
	off_t fat_offset;
	size_t fat_size;
	off_t end;
	size_t size;
	off_t root_dir;
};

struct fatx_internal_file_record {
	uint8_t name_length;
	uint8_t attributes;
	uint8_t name[42];
	uint32_t first_cluster;
	uint32_t size;
	uint32_t modified_time;
	uint32_t created_time;
	uint32_t accessed_time;
}__attribute__((packed));

#define fatx_warn_corruption(fmt, ...) fprintf(stderr, "libfatx: Warning: Possible filesystem corruption:\n" fmt "\n(file: %s, line: %d)\n", __VA_ARGS__, __FILE__, __LINE__)

/**
 * Converts a name from a FATX record into ansi format (null-terminated).
 * A FATX name cannot be more than 42 characters, which means an ansi_name
 * allocated to 43 characters will fit any FATX name.
 */
static void fatx_name_fatx2ansi(char *ansi_name, uint8_t *fatx_name, int length) {
	int i;
	if (length > 42) length = 42;
	for (i = 0; i < length; i++) {
		ansi_name[i] = fatx_name[i];
	}
	ansi_name[length] = 0;
}

/**
 * Converts a name from ansi format into a FATX record (followed by 42-len FF bytes)
 * A FATX name cannot be more than 42 characters, which means that excess characters
 * in ansi_name will be dropped. The length field in a FATX file record will need to
 * be updated.
 */
static void fatx_name_ansi2fatx(uint8_t *fatx_name, char *ansi_name, int *length) {
	int i;
	*length = strlen(ansi_name);
	if ((unsigned int)(*length) > 42) *length = 42;
	for (i = 0; i < *length; i++) {
		fatx_name[i] = ansi_name[i];
	}
	for (; i < 42; i++) {
		fatx_name[i] = 0xFF;
	}
}

static time_t fatx_time_fatx2unix(uint32_t time) {
	struct tm t;
	t.tm_year = (time >> 25) + 80;
	t.tm_mon = ((time >> 21) & 0xF) - 1;
	t.tm_mday = (time >> 16) & 0x1F;
	t.tm_hour = (time >> 11) & 0x1F;
	t.tm_min = (time >> 5) & 0x3F;
	t.tm_sec = time & 0x1F;
	return mktime(&t);
}

static uint32_t fatx_time_unix2fatx(time_t time) {
	uint32_t ret;
	struct tm t;
	ret = 0;
	if (localtime_r(&time, &t) != &t) {
		fprintf(stderr, "libfatx: Warning: localtime_r could not convert the time.");
		return -1;
	}
	ret |= (t.tm_year - 80) << 25;
	ret |= ((t.tm_mon + 1) & 0xF) << 21;
	ret |= (t.tm_mday & 0x1F) << 16;
	ret |= (t.tm_hour & 0x1F) << 11;
	ret |= (t.tm_min & 0x3F) << 5;
	ret |= t.tm_sec & 0x1F;
	return ret;
}

static inline uint32_t fatx_read_int(fatx_fs_info *info, off_t offset);
static inline uint16_t fatx_read_short(fatx_fs_info *info, off_t offset);

static inline off_t fatx_next_cluster_offset(fatx_fs_info *info, off_t offset);

int fatx_find_file_offsets(fatx_file_offsets *offsets,
		fatx_fs_info *info, const char *path) {
	char *d_path, *save_ptr, *token, name[43];
	int i;
	off_t record_offset, data_offset;
	struct fatx_internal_file_record records[256];
	if (!strcmp(delimiter, path)) { // root directory, has no record
		offsets->record_offset = -1;
		offsets->data_offset = info->root_dir;
		return 0;
	}
	d_path = strdupa(path);
	data_offset = info->root_dir;
	token = strtok_r(d_path, delimiter, &save_ptr);
	if (token == NULL) return -1;
	do {
		while (1) {
			size_t read = pread(info->fd, records, sizeof(records), data_offset);
			for (i = 0; i < 256; i++) {
				if (records[i].name_length == 0xFF) {
					return -1;
				} else if (records[i].name_length == 0xE5) { // deleted file, skip
					continue;
				} else if (records[i].name_length > 42) {
					fatx_warn_corruption("Filename length is an invalid value (possible that we stepped into a file somehow)\nname_length: %d", records[i].name_length);
					return -1;
				}
				fatx_name_fatx2ansi(name, records[i].name, records[i].name_length);
				if ((strlen(token) == records[i].name_length) && (!strcasecmp(token, name))) {
					record_offset = (off_t) (sizeof(struct fatx_internal_file_record) * i) + data_offset;
					if (info->endianness == BIG_ENDIAN) data_offset = be32toh(records[i].first_cluster) - 1;
					if (info->endianness == LITTLE_ENDIAN) data_offset = le32toh(records[i].first_cluster) - 1;
					data_offset <<= 14;
					data_offset += info->root_dir;
					break;
				}
			}
			if ((strlen(token) == records[i].name_length) && (!strcasecmp(token, name))) {
				break;
			}
			data_offset = fatx_next_cluster_offset(info, data_offset);
			if (data_offset < 0) return -1;
		}
	} while ((token = strtok_r(NULL, delimiter, &save_ptr)) != NULL);
	offsets->record_offset = record_offset;
	offsets->data_offset = data_offset; // you can see that there's problems if sizeof(off_t) < 64 bits
	return 0;
}

int fatx_read_file_record(fatx_file_record *file_record,
		fatx_fs_info *info, const char *path) {
	fatx_file_offsets offsets;
	memset(file_record, 0, sizeof(fatx_file_record));
	int ret = fatx_find_file_offsets(&offsets, info, path);
	if (ret < 0) return ret;
	struct fatx_internal_file_record internal_file_record;
	if (offsets.record_offset == -1 && offsets.data_offset == info->root_dir) {
		file_record->name[0] = '/';
		file_record->name[1] = 0;
		file_record->isdir = 1;
		return 0;
	}
	size_t read = pread(info->fd, &internal_file_record,
			sizeof(struct fatx_internal_file_record), offsets.record_offset);
	if (read != sizeof(struct fatx_internal_file_record)) { return -1; }
	fatx_name_fatx2ansi(file_record->name, internal_file_record.name,
			internal_file_record.name_length);
	if (internal_file_record.attributes & 0x10) {
		file_record->isdir = 1;
	} else {
		file_record->isdir = 0;
	}
	if (info->endianness == LITTLE_ENDIAN) {
		file_record->size = (size_t) le32toh(internal_file_record.size);
		file_record->modified = fatx_time_fatx2unix(le32toh(internal_file_record.modified_time));
		file_record->created = fatx_time_fatx2unix(le32toh(internal_file_record.created_time));
		file_record->accessed = fatx_time_fatx2unix(le32toh(internal_file_record.accessed_time));
	} else if (info->endianness == BIG_ENDIAN) {
		file_record->size = (size_t) be32toh(internal_file_record.size);
		file_record->modified = fatx_time_fatx2unix(be32toh(internal_file_record.modified_time));
		file_record->created = fatx_time_fatx2unix(be32toh(internal_file_record.created_time));
		file_record->accessed = fatx_time_fatx2unix(be32toh(internal_file_record.accessed_time));
	} else {
		return -1;
	}
	return 0;
}

/**
 * Reads a short from a given offset without moving the file marker.
 */
static inline uint16_t fatx_read_short_fd(int fd, off_t offset) {
	uint16_t ret;
	size_t read = pread(fd, &ret, sizeof(uint16_t), offset);
	return ret;
}

/**
 * Reads a short from a given offset without moving the file marker.
 * Takes into account the endianness of the filesystem.
 */
static inline uint16_t fatx_read_short(fatx_fs_info *info, off_t offset) {
	uint16_t ret;
	size_t read = pread(info->fd, &ret, sizeof(uint16_t), offset);
	if (info->endianness == BIG_ENDIAN) ret = be32toh(ret);
	if (info->endianness == LITTLE_ENDIAN) ret = le32toh(ret);
	return ret;
}

/**
 * Reads an integer from a given offset without moving the file marker.
 */
static inline uint32_t fatx_read_int_fd(int fd, off_t offset) {
	uint32_t ret;
	size_t read = pread(fd, &ret, sizeof(uint32_t), offset);
	return ret;
}

/**
 * Reads an integer from a given offset without moving the file marker.
 * Takes into account the endianness of the filesystem.
 */
static inline uint32_t fatx_read_int(fatx_fs_info *info, off_t offset) {
	uint32_t ret;
	size_t read = pread(info->fd, &ret, sizeof(uint32_t), offset);
	if (info->endianness == BIG_ENDIAN) ret = be32toh(ret);
	if (info->endianness == LITTLE_ENDIAN) ret = le32toh(ret);
	return ret;
}

/**
 * FATX doesn't have any information in the header realating to the size of the
 * filesystem or the location of the root directory. Instead, this information is
 * calculated based off of the size of the partition/drive.
 */
static inline void fatx_calc_size_and_table_offset(fatx_fs_info *info) {
	off_t here = lseek(info->fd, 0, SEEK_CUR);
	info->fat_offset = (off_t) 0x1000;
	info->end = lseek(info->fd, 0, SEEK_END);
	info->width = (info->end < 0x3FFF4000) ? sizeof(uint16_t) : sizeof(uint32_t);
	if (info->width == sizeof(uint32_t)) {
		info->root_dir = -(-((info->end >> 12) + 1) & INT64_C(-0x1000)) + info->fat_offset;
	} else if (info->width == sizeof(uint16_t)) {
		info->root_dir = -(-((info->end >> 13) + 1) & INT64_C(-0x1000)) + info->fat_offset;
	}
	info->size = info->end - info->root_dir;
	info->fat_size = info->size >> 14;
	lseek(info->fd, here, SEEK_SET);
}

/**
 * Find endianness of the filesystem pointed to by filename.
 * Returns LITTLE_ENDIAN for little endian, BIG_ENDIAN for big endian,
 * or -1 if it is not a fatx partition. (BIG_ENDIAN and LITTLE_ENDIAN
 * are defined in endian.h)
 */
static inline int fatx_find_endianness(int fd) {
	uint32_t magic = fatx_read_int_fd(fd, 0);
	/*
	 * Little endian partitions have FATX as their magic identifier.
	 * Since the code was never changed, big endian partitions have XTAF.
	 * Note that this is backwards from what it seems it should be.
	 */
	if (be32toh(magic) == FATX_MAGIC) { // Ascii: F A T X
		return LITTLE_ENDIAN;
	} else if (le32toh(magic) == FATX_MAGIC) { // Ascii: X T A F
		return BIG_ENDIAN;
	}
	return -1;
}

static uint32_t fatx_next_cluster(fatx_fs_info *info, uint32_t cluster);

static inline off_t fatx_next_cluster_offset(fatx_fs_info *info, off_t offset) {
	off_t ret;
	ret = (off_t)(fatx_next_cluster(info, ((offset - info->root_dir) >> 14) + 1));
	if (ret == 0xFFFFFFFF) return -1;
	if (ret == 0xFFFFFFFE) return -2;
	return ((off_t)(ret - 1) << 14) + info->root_dir;
}

static uint32_t fatx_next_cluster(fatx_fs_info *info, uint32_t cluster) {
	if (cluster == 1) return -1; // root directory can only be one cluster
	switch(info->width) {
	case sizeof(uint32_t):
		if ((cluster & 0xFFFFFFF) > info->fat_size && (cluster & 0xFFFFFFF) <  0xFFFFFF5) {
			fatx_warn_corruption("Current cluster is out of bounds\ncurrent_cluster: %d", cluster);
			return -1;
		}
		if ((cluster & 0xFFFFFFF) > 0xFFFFFF5) {
			fprintf(stderr, "libfatx: Warning: fatx_next_cluster was given an invalid cluster.\n");
			return -1;
		}
		cluster = fatx_read_int(info, info->fat_offset + cluster * sizeof(uint32_t));
		if ((cluster & 0xFFFFFFF) > info->fat_size && (cluster & 0xFFFFFFF) <  0xFFFFFF5) {
			fatx_warn_corruption("Current cluster is out of bounds\ncurrent_cluster: %d", cluster);
			return -1;
		}
		if ((cluster & 0xFFFFFFF) > 0xFFFFFF5) { // last cluster
			return -2;
		}
		return cluster;
	case sizeof(uint16_t):
		if (cluster > info->fat_size && cluster <  0xFFF5) {
			fatx_warn_corruption("Current cluster is out of bounds\ncurrent_cluster: %d", cluster);
			return -1;
		}
		if (cluster > 0xFFF5) {
			fprintf(stderr, "libfatx: Warning: fatx_next_cluster was given an invalid cluster.\n");
			return -1;
		}
		cluster = fatx_read_int(info, info->fat_offset + cluster * sizeof(uint32_t));
		if (cluster > info->fat_size && cluster <  0xFFF5) {
			fatx_warn_corruption("Current cluster is out of bounds\ncurrent_cluster: %d", cluster);
			return -1;
		}
		if (cluster > 0xFFF5) { // last cluster
			return -1;
		}
		return cluster;
	}
}

/**
 * Opens the file pointed to by filename. Returns a pointer to data
 * for the other fatx functions to use.
 */
fatx_fs_info *fatx_fs_init(const char *filename) {
	int fd, endianness;
	fatx_fs_info *info;
	info = malloc(sizeof(fatx_fs_info));
	if (info == NULL) {
		fputs("libfatx: fatal: Out of memory\n", stderr);
		return NULL;
	}
	fd = open(filename, O_RDWR);
	info->mode = O_RDWR;
	if (fd < 0) {
		if (errno == EACCES) { // permission denied; let's try read only
			fd = open(filename, O_RDONLY);
			if (fd < 0) {
				fprintf(stderr, "libfatx: Error opening file %s: [%d] %s\n",
						filename, errno, strerror(errno));
				return NULL;
			} else {
				fprintf(stderr, "libfatx: Warning: Opened file %s in read-only mode\n",
						filename);
				info->mode = O_RDONLY;
			}
		} else {
			fprintf(stderr, "libfatx: Error opening file %s: [%d] %s\n", filename,
					errno, strerror(errno));
			return NULL;
		}
	}
	info->fd = fd;
	endianness = fatx_find_endianness(fd);
	if (endianness < 0) {
		fprintf(stderr, "libfatx: Error: %s is not a FATX filesystem\n",
				filename);
		return NULL;
	}
	info->endianness = endianness;
	fatx_calc_size_and_table_offset(info);
	return info;
}

/**
 * Calls func on each file in the (sub)directory
 * Think of it as an "ls" type function
 */
int fatx_list_dir(fatx_fs_info *info, const char *path, void (*func)(const char *, void *), void *user) {
	off_t data_offset;
	fatx_file_offsets offsets;
	struct fatx_internal_file_record ifr;
	int ret = fatx_find_file_offsets(&offsets, info, path);
	if (ret < 0) return ret;
	if (offsets.record_offset > 0 || offsets.data_offset != info->root_dir) { // to avoid ENOTDIR on root directory
		pread(info->fd, &ifr, sizeof(ifr), offsets.record_offset);
		if ((ifr.attributes & 0x10) != 0x10) return -ENOTDIR;
	}
	data_offset = offsets.data_offset;
	while (1) {
		struct fatx_internal_file_record records[256];
		size_t read = pread(info->fd, records, sizeof(records), data_offset);
		int i;
		for (i = 0; i < 256; i++) {
			if (records[i].name_length == 0xFF) {
				return 0;
			} else if (records[i].name_length == 0xE5) { // deleted file, skip
				continue;
			} else if (records[i].name_length > 42) {
				fatx_warn_corruption("Filename length is an invalid value (possible that we stepped into a file somehow)\nname_length: %d", records[i].name_length);
				return -1;
			}
			char name[43];
			fatx_name_fatx2ansi(name, records[i].name, records[i].name_length);
			func(name, user);
		}
		data_offset = fatx_next_cluster_offset(info, data_offset);
		if (data_offset == -2) return 0;
		else if (data_offset < 0) return -1;
	}
	return 0;
}

/**
 * Reads size bytes from a file starting at offset. Expects buffer to be allocated for at least size bytes.
 */
size_t fatx_read_file(fatx_fs_info *info, const char *path, void *buffer, size_t size, off_t offset) {
	int i;
	uint32_t cluster;
	off_t data_offset;
	size_t read;
	fatx_file_offsets offsets;
	struct fatx_internal_file_record ifr;
	int ret = fatx_find_file_offsets(&offsets, info, path);
	if (ret < 0) return ret;
	if (offsets.record_offset < 0 && offsets.data_offset == info->root_dir) { // to avoid ENOTDIR on root directory
		return EISDIR;
	}
	pread(info->fd, &ifr, sizeof(struct fatx_internal_file_record), offsets.record_offset);
	if (ifr.attributes & 0x10) return EISDIR;
	switch(info->endianness) {
	case BIG_ENDIAN:
		cluster = be32toh(ifr.first_cluster);
		break;
	case LITTLE_ENDIAN:
		cluster = le32toh(ifr.first_cluster);
		break;
	default:
		return -1;
	}
	data_offset = offsets.data_offset;
	for (i = 0; i < (offset >> 14); i++) {
		data_offset = fatx_next_cluster_offset(info, data_offset);
		if (data_offset < 0) return -1;
	}
	offset -= (off_t)(i) << 14;
	read = 0;
	for (i = 0; i < ((offset + size) >> 14); i++) {
		read += pread(info->fd, (uint8_t *)(buffer) + read, 0x4000 - offset, data_offset + offset);
		offset = 0;
		data_offset = fatx_next_cluster_offset(info, data_offset);
		if (data_offset < 0) return -1;
	}
	size -= (off_t)(i) << 14;
	read += pread(info->fd, (uint8_t *)(buffer) + read, size, data_offset + offset);
	return read;
}

void fatx_fs_end(fatx_fs_info *info) {
	close(info->fd);
	free(info);
}
